nym-vpn-x/src-tauri/src/error.rs

```diff
@@ -20,100 +24,160 @@ pub enum CmdErrorSource {
+    pub key: ErrorKey,
+    /// Extra data to be passed along to help specialize the problem
+    pub data: Option<HashMap<String, String>>,
 }
 
-impl CmdError {
-    pub fn new(error: CmdErrorSource, message: &str) -> Self {
+impl BkdError {
+    pub fn new(message: &str, key: ErrorKey) -> Self {
+        Self {
+            message: message.to_string(),
+            key,
+            data: None,
+        }
+    }
+
+    pub fn new_with_data(
+        message: &str,
+        key: ErrorKey,
+        data: Option<HashMap<String, String>>,
+    ) -> Self {
         Self {
             message: message.to_string(),
-            source: error,
-            i18n_key: None,
+            key,
+            data,
         }
     }
 
-    pub fn new_with_local(error: CmdErrorSource, message: &str, key: I18nKey) -> Self {
+    pub fn new_internal(message: &str, data: Option<HashMap<String, String>>) -> Self {
         Self {
             message: message.to_string(),
-            source: error,
-            i18n_key: Some(key),
+            key: ErrorKey::InternalError,
+            data,
         }
     }
 }
 
-impl Display for CmdError {
+impl Display for BkdError {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{}: {}", self.source, self.message)
+        write!(
+            f,
+            "message '{}' key [{:?}] data [{:?}]",
+            self.message,
+            self.key,
+            self.data.as_ref()
+        )
     }
 }
 
-impl From<VpndError> for CmdError {
+impl From<VpndError> for BkdError {
     fn from(error: VpndError) -> Self {
         match error {
-            VpndError::RpcError(s) => CmdError::new(
-                CmdErrorSource::DaemonError,
-                &format!("failed to call the daemon: {}", s),
-            ),
-            VpndError::FailedToConnectIpc(_) | VpndError::FailedToConnectHttp(_) => {
-                CmdError::new(CmdErrorSource::DaemonError, "not connected to the daemon")
+            VpndError::GrpcError(s) => {
+                BkdError::new(&format!("grpc error: {}", s), ErrorKey::GrpcError)
             }
+            VpndError::FailedToConnectIpc(_) | VpndError::FailedToConnectHttp(_) => BkdError::new(
+                "not connected to the daemon",
+                ErrorKey::NotConnectedToDaemon,
+            ),
+        }
+    }
+}
+
+impl From<nym_vpn_proto::Error> for BkdError {
+    fn from(error: nym_vpn_proto::Error) -> Self {
+        Self {
+            message: error.message.clone(),
+            key: error.kind().into(),
+            data: error.details.into(),
         }
     }
 }
 
-/// Enum of the localization keys for error messages
-/// displayed in the UI
-#[derive(Debug, Serialize, Deserialize, TS)]
+/// Enum of the possible specialized errors emitted by the daemon
+/// or the app backend side, to be passed to the UI layer
+#[derive(Debug, Serialize, Deserialize, TS, Clone)]
 #[ts(export)]
-pub enum I18nKey {
+pub enum ErrorKey {
```

Comment:
It looks like you're merging 3 kinds of errors into one global one, is there a frontend reason for this? It looks to me like you want to keep them separate, especially the credential and connection errors since they can never happen in the same situations. Also it looks like this information is still weakly encoded in the prefix-name?

Reply:
Yes it is, the whole purpose of this cross error handling is to expose to the UI layer a unique set of explicitly defined errors and which are the only possible errors that can happens and having impact at the UI layer.
Enforcing that allows the frontend to create whatever system it needs on top of it. Eg. basing the localization of error messages.
Exposing separated errors types to the frontend ui instead of one unique type, just add more complexity for no benefits.

> Also it looks like this information is still weakly encoded in the prefix-name?

I'm not sure why you said "weakly" but since a key in that enum can represent a very specific error case, the name has to be unique and self-descriptive enough (+ rust doc added details for each variants).

Reply:
By "weakly" here I mean weak typing, since it looks like in the error enum that you are merging to disjoint error types, but still separating by prefixing them in the name with "Connection" and "Credential"

> Yes it is, the whole purpose of this cross error handling is to expose to the UI layer a unique set of explicitly defined errors and which are the only possible errors that can happens and having impact at the UI layer.

So that's the thing, by merging them means that you can't look at the error enum and tell exactly which errors that can happen (e.g. you can't have credential import errors when connecting). I don't know how your are consuming these, but I advise you to double check this to confirm you really want a global error like this

Reply:
Let's discuss it internally please